src/control/server/storage/config.go

```diff
@@ -241,6 +251,94 @@ func (tcs TierConfigs) Validate() error {
 	return nil
 }
 
+func (tcs TierConfigs) validateBdevTierRoles() error {
+	sc := tcs.ScmConfigs()[0]
+	bcs := tcs.BdevConfigs()
+
+	nrWALTiers := 1
+	nrIndexTiers := 1
+	if sc.Class == ClassDcpm {
+		nrWALTiers = 0
+		nrIndexTiers = 0
+		if len(bcs) > 1 {
+			return errors.Errorf("more than 1 bdev tiers not allowed with dcpm scm class")
+		}
+	}
+
+	var foundWALTiers, foundIndexTiers, foundDataTiers int
+	for _, bc := range bcs {
+		bits := bc.Bdev.DeviceRoles.OptionBits
+		if (bits & BdevRoleWAL) != 0 {
+			foundWALTiers++
+		}
+		if (bits & BdevRoleIndex) != 0 {
+			foundIndexTiers++
+		}
+		if (bits & BdevRoleData) != 0 {
+			foundDataTiers++
+		}
+	}
+
+	if foundWALTiers != nrWALTiers {
+		return errors.Errorf("found %d WAL tiers, wanted %d", foundWALTiers, nrWALTiers)
+	}
+	if foundIndexTiers != nrIndexTiers {
+		return errors.Errorf("found %d Index tiers, wanted %d", foundIndexTiers, nrIndexTiers)
+	}
+	// When bdev NVMe tiers exist, there should always be at least one Data tier.
+	if foundDataTiers == 0 {
+		return errors.New("found 0 Data tiers, wanted at least 1")
+	}
+
+	return nil
+}
+
+// Set NVME class tier roles either based on explicit settings or heuristics.
+func (tcs TierConfigs) assignBdevTierRoles() error {
+	scs := tcs.ScmConfigs()
+	bcs := tcs.BdevConfigs()
+
+	// Require tier-0 to be a SCM tier.
+	if len(scs) != 1 || scs[0].Tier != 0 {
+		return errors.New("first storage tier is not scm")
+	}
+
+	// Skip role assignment and validation no NVMe tiers exist.
+	if !tcs.HaveRealNVMe() {
+		return nil
+	}
+
+	// Skip implicit role assignment if any roles have already been explicitly assigned.
+	for _, bc := range bcs {
+		if !bc.Bdev.DeviceRoles.IsEmpty() {
+			return tcs.validateBdevTierRoles()
+		}
+	}
+
```

Comment:
There is an all or nothing (in terms of implicit assignment) assumption here which I don't think is unreasonable. As we discussed on a call extra validation should be added when we see a need for it after real usage. The main requirements are that the rules do not get duplicated over multiple tiers as per the current validation, the engine is less concerned about the tier orders after roles have been assigned.

Reply:
Sure, I think it's fine here to assume that if any tiers have a role set, we can short-circuit the rest of the implicit role assignment logic. My point though is that it seems like it would be better for UX if the validation code at least ensured that every tier has a role set, even if we are still figuring out the bigger picture of what constitutes a valid configuration. I could see people being confused/frustrated if they have N tiers defined but only a subset are used because they forgot to define roles on every tier.